src/main/scala/com/fulcrumgenomics/commons/io/CommandLineTool.scala

```diff
@@ -0,0 +1,162 @@
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package com.fulcrumgenomics.commons.io
+
+import java.nio.file.{Files, Path}
+
+import com.fulcrumgenomics.commons.util.LazyLogging
+
+import scala.collection.mutable
+import scala.io.Source
+import scala.util.{Failure, Success, Try}
+
+trait CommandLineTool extends LazyLogging {
+  /** The name of the executable such as Rscript or gs. */
+  val Executable: String
+
+  /** Command used for the above executable. */
+  val TestCommand: Seq[String]
+
+  /** Exception class that holds onto an executable's exit/status code. */
+  case class ToolException(status: Int) extends RuntimeException {
+    override def getMessage: String = s"$Executable failed with exit code $status."
+  }
+
+  /** Returns a tuple of (Boolean, String): first element is true when command can execute
+    * with no error; second element is a string of command stdout */
+  def execArgs(command: String*): (Boolean, String) = {
+    try {
+      // Read redirected stdout to a string
+      val process = new ProcessBuilder(command: _*).redirectErrorStream(true).start()
+      val stdout  = mutable.ListBuffer[String]()
+      val sink    = new AsyncStreamSink(process.getInputStream, s => stdout.append(s))
+      // True if the command can execute with no error
+      val canExecute = process.waitFor() == 0
+      sink.close()
+
+      (canExecute, stdout.toList.mkString)
+    }
+    catch {case e: Exception => (false, "")}
+  }
+
+  /** Returns true if the tool is available and false otherwise. */
+  lazy val Available: Boolean = {
+    execArgs(Executable +: TestCommand: _*)._1
+  }
+}
+
+trait CanRunScript {
+  self: CommandLineTool =>
+  /** Suffix for scripts that can be run by this tool. */
+  val Suffix: String
+
+  /** Executes a script from the classpath if the tested executable is available. */
+  def execIfAvailable(scriptResource: String, args: String*): Unit =
+    try {
+      if (Available) exec(scriptResource, args:_*)
+    } catch {case e: Exception => throw new Exception(s"Cannot execute script from $scriptResource with args:$args") }
+
+  /** Executes from a script stored at a Path if the tested executable is available. */
```

Comment:
Use the imperative mood:

```scala
/** Execute a script if this executable is available, raise an exception otherwise. */
```

Here and throughout

Reply:
👍 BTW just to be clear.. we aren't raising exception at this line for [`execIfAvailable`](https://github.com/fangylo/commons/blob/c06e2da1ea0acfc291914af77e47947600c78856/src/main/scala/com/fulcrumgenomics/commons/io/CommandLineTool.scala#L77-L82) if the executable is *not* available.